_posts/2016-08-03-better-topic-detection-in-text-with-lda2vec.md

```diff
@@ -0,0 +1,106 @@
+
+All this is trained with feedforward neural networks (networks without cycles, straightforward input – projection – hidden – output layers).
+
+This is a computationally intensive process, so much work has been done in getting clever shortcuts.
+
+Improvements that were added in the model: using idiomatic phrases instead of words – some words should be grouped together, say the Financial Times has little to do with time measurement.
+
+As an interesting aside: Google has the software patent for this technique, but doesn't seem to have done much with that patent at the moment.
+
+## Lda2vec
+
+Lda2vec is a research project by Chris E. Moody, PhD at Caltech.  Lda2vec’s aim is to find topics while also learning word vectors to obtain sparser topic vectors that are easier to interpret, while also training the other words of the topic in the same vector space (using neighbouring words).
+
+This is achieved by changing the objective function of skip-gram negative sampling we described in previous paragraph to add document feature vectors in the mix. The intuition here is that a word is enriched by the knowledge of the document it’s embedded in: if a document is about airlines, Germany might be close to Lufthansa, while another document might evoke other associations.
+
+The desired result is both words and document vectors, where word vectors can be added to the document vectors for meaningful results.
+
+#### Training
+
+The objective function is trained using stochastic gradient descent to look for minima (Adam optimizer).
+
+At every iteration, the metric to see if we’ve achieved the right result is done by trying to find a coherence measure (with <a href="http://palmetto.aksw.org/palmetto-webapp/">Palmetto</a>, an online tool). This examines a measure over large sliding windows in the documents and checks whether the calculated probabilities of occuring for neighbouring pairs of words looks high (NPMI normalized pointwise mutual information).
+
+#### Results
+
+This technique was used on a corpus of bulletin board messages, and on Hacker News comments, and yielded coherent topics.
+
+| Topic label        | "Space"           | "Encryption"  |"X-Windows"|"Middle East"
+| ------------- |-------------| -----|
+| Top tokens     | astronomical<br>Astronomy<br>satellite<br>planetary | encryption<br>encrypt<br>wiretap<br>escrow |mydisplay<br>xlib<br>window<br>cursor|Armenian<br>Lebanese<br>Muslim<br>Turk
+| Topic coherence      | 0.712      | 0.675 |0.472|0.615
+
+
+
+## Conclusion
+
+This is a research project – exceptionally, it has really decent open source code in Python which is rare for research papers (props to Chris Moody). The technique looks promising, and intuitively makes sense, and the results look exciting.
+
+Running time: on a high-spec laptop it takes multiple-days-scale scale durations to run the bulletin board example (11312 documents, 3.4GB, 20 topics) – this may be shorter when using the CUDA optimizations he uses.
```

Comment:
.. and there's definitely room to parallellize the calculations to obtain better running times.